### DISCUSSION TITLE
CP-Cluster is improving Yolov5 accuracy without retraining official models

### DISCUSSION THREAD

**Discussion by shenyi0220** [ultralytics/yolov5#6227] [Show and tell]
CP-Cluster outperforms all NMS methods when applied to Yolov5 official models without getting them retrained, gaining ~0.3 improvement on average.
https://github.com/shenyi0220/cp-cluster

**Comment by glenn-jocher** [comment]
@shenyi0220 yes I'd seen that! Those are impressive mAP improvements. What's the best way to offer this an an NMS option? Our NMS function is here, it's mainly a wrapper around the torch vision nms function, which itself is compiled in C I think:

https://github.com/ultralytics/yolov5/blob/f80c463010101e463670440537662f671cbaaa04/utils/general.py#L647-L669

**Reply by shenyi0220** [reply]
In my exp impl, I directly replaced the torchvision.ops.nms with the mmcv-based API of CP-Cluster.

So later for yolov5, when I got green light from the NV to fully open source GPU versions of this method, maybe I can contribute like this:
-Opt1: upstreamed the CP-Cluster cpu/gpu versions to mmcv, and support optional call of mmcv API in addition to torchvision nms.
-Opt2: directly implemented gpu versions in yolov5 repo, while this may need optional compile step
-Opt3: try to upstream the gpu version of cp-cluster to torchvision

**Reply by zhiqwang** [reply]
Hi @shenyi0220 , Do you think it is possible to make this as a separate package?

**Reply by glenn-jocher** [reply]
@shenyi0220 opt3 would be awesome! But I know the torch team doesn't like to include new development very fast. It took them a very long time to add the Swish activation for example.

We can't support any compilation steps in YOLOv5, so then Opt1 might be best while you work on Opt3. For Opt3 please raise an issue on the torch repo and @ me and I'll throw my support behind it.
